fsgrid.hpp

```diff
@@ -1040,3 +1086,108 @@ template <typename T, int stencil> class FsGrid : public FsGridTools{
+   std::array<T, N> res;
+    for (size_t i = 0; i < N; ++i){
+        res[i] = arr1[i] + arr2[i];
+      }
+    return res; 
+}
+
+template <typename T, size_t N>
+static inline std::array<T,N> operator-(const std::array<T, N>& arr1, const std::array<T, N>& arr2){
+   std::array<T, N> res;
+    for (size_t i = 0; i < N; ++i){
+        res[i] = arr1[i] - arr2[i];
+      }
+    return res; 
+}
+
+template <typename T, size_t N>
+static inline std::array<T, N> operator*(const std::array<T, N> &arr1, const std::array<T, N> &arr2){
+   std::array<T, N> res;
+   for (size_t i = 0; i < N; ++i){
+      res[i] = arr1[i] * arr2[i];
+   }
+   return res;
+}
+
+template <typename T, size_t N>
+static inline std::array<T,N> operator/(const std::array<T, N>& arr1, const std::array<T, N>& arr2){
+   std::array<T, N> res;
+    for (size_t i = 0; i < N; ++i){
+        res[i] = arr1[i] / arr2[i];
+      }
+    return res; 
+}
+
+template <typename T, size_t N,typename S,typename std::enable_if<std::is_arithmetic<S>::value>::type* = nullptr>
+std::array<T,N> operator*(std::array<T, N>& arr, S mult){
+    for (size_t i = 0; i < N; ++i)
+        arr[i]*=mult;
+    return arr; 
+}
+
+
+
+// FsGrid Operator Overloads
+template <class T, int stencil>
+static inline FsGrid<T, stencil> operator+(const FsGrid<T, stencil> &lhs, const FsGrid<T, stencil> &rhs){
+   FsGrid<T, stencil> a(lhs);
+   return a+=rhs;
+}
+
+template <class T, int stencil>
+static inline FsGrid<T, stencil> operator-(const FsGrid<T, stencil> &lhs, const FsGrid<T, stencil> &rhs){
+   FsGrid<T, stencil> a(lhs);
+   return a-=rhs;
+}
+
+template <class T, int stencil>
+static inline FsGrid<T, stencil> operator*(const FsGrid<T, stencil> &lhs, const FsGrid<T, stencil> &rhs){
+   FsGrid<T, stencil> a(lhs);
+   return a *= rhs;
+}
+
+template <class T, int stencil>
+static inline FsGrid<T, stencil> operator/(const FsGrid<T, stencil> &lhs, const FsGrid<T, stencil> &rhs){
+   FsGrid<T, stencil> a(lhs);
+   return a /= rhs;
+}
+
+template <class T, int stencil,typename S,typename std::enable_if<std::is_arithmetic<S>::value>::type* = nullptr>
+static inline FsGrid<T, stencil>  lerp_t(const  FsGrid<T, stencil> &lhs, const FsGrid<T, stencil> &rhs,S t0, S t1,S t){
+   
+   //Sanity check for timestamps. If sizes mismatch then this will make it crash in the '=' operator
+   bool doAble = (t0<=t) && (t<=t1);
+   if (!doAble){
+      #pragma message(": warning<FsGrid lerp_t>: We should probably crash here instead of returning nonesense")
+      //Do we crash here?
+      std::cerr<<"Interpolation not doable in the range of "<<t0<<" "<<t<<" "<<t1<<std::endl; 
+      return lhs;
+   }
+
+   //First let's create the targer grid by copying one of the inputs
+   FsGrid<T, stencil> targetGrid(lhs);
+
+   //Let's now calculate the intepolation weights;
+   S range=t1-t0;
+   S d0= abs(t-t0);
+   S d1= abs(t-t1);
+   S w0= 1.- (d0/range);
+   S w1= 1.- (d1/range);
```

Comment:
Does this need a division-by-zero protection? (Realistically, if t1==t0, we should just return lhs here, I suppose)

Reply:
I was thinking about it..True fixing it

Reply:
@ursg  I am changing  this now but with floating point t0,t1 the == operator does not make much sense at times. Should I instead use a tolerance under which we assume t1==t0?

Reply:
Realistically, the scenario that I'm thinking about is that *someone* (I'm looking at you, absorbing boundaries), iterates through N time-lagged copies of an fsgrid and somehow calculates ```lerp(current_state, previous_state[n], some_time)```. Then there might be edge cases (such as restarts), where previous_state[n] happens to be at the same time as current_state. In this case, I would prefer it to... not crash.

Reply:
So there, exact floating point comparison with == actually does make sense.